Library/MeterGeometry.cpp

```diff
@@ -0,0 +1,370 @@
+								break;
+						}
+						else
+							postOptions[shapeOption.c_str()] = std::pair<LPCWSTR, LPCWSTR>(optionName.c_str(), optionValue.c_str());
+					}
+					else
+					{
+						
+						shapeType = optionName;
+						shapeParameters = optionValue;
+					}
+				}
+				else
+					m_MeasureOptions[shapeName.c_str()].push_back(std::pair<std::wstring, std::wstring>(optionName, optionValue));
+				++optionId;
+			}
+			if (shapeType.empty() || shapeParameters.empty())
+				break; //Invalid shape
+			if (!ParseShape(m_Shapes[shapeName.c_str()], shapeType.c_str(), shapeParameters.c_str()))
+				break;
+			shapeName = L"Shape" + std::to_wstring(++currentShapeId);
+			shapeOption = parser.ReadString(section, shapeName.c_str(), L"");
+		} while (!shapeOption.empty());
+
+	for (const auto& option : postOptions)
+	{
+		if (!ReplaceShapeOption(m_Shapes[option.first], option.second.first, option.second.second))
+			continue;
+	}
+	if (m_Initialized && m_Measures.empty() && !m_DynamicVariables)
+	{
+		Initialize();
+		m_NeedsRedraw = true;
+	}
+}
+
+bool MeterGeometry::Update()
+{
+	if (Meter::Update())
+	{
+		if (!m_Measures.empty() || m_DynamicVariables)
+		{
+			for (const auto& shape : m_MeasureOptions)
+			{
+				for (const auto& option : shape.second) {
+					std::wstring optionValue = option.second;
+					bool replaced = ReplaceMeasures(optionValue, AUTOSCALE::AUTOSCALE_OFF, 1.0, 6, false);
+					if (replaced)
+						if (!IsShape(option.first.c_str())) {
+							if (!ReplaceShapeOption(m_Shapes[shape.first], option.first.c_str(), optionValue.c_str()))
+								continue;
+						}
+						else
+							if (!ParseShape(m_Shapes[shape.first], option.first.c_str(), optionValue.c_str()))
+								continue;
+				}
+			}
+			return true;
+		}
+		else if (m_NeedsRedraw)
+		{
+			m_NeedsRedraw = false;
+			return true;
+		}
+	}
+	return false;
+}
+
+bool MeterGeometry::Draw(Gfx::Canvas & canvas)
+{
+	if (!Meter::Draw(canvas)) return false;
+	
+	for (const auto& shape : m_Shapes) {
+		if (shape.second.m_Shape) {
+			D2D1_POINT_2F centerPoint;
+			centerPoint = D2D1::Point2F(shape.second.m_RotationCenter.x, shape.second.m_RotationCenter.y);
+			D2D1_MATRIX_3X2_F transform = D2D1::Matrix3x2F(
+				D2D1::Matrix3x2F::Rotation(shape.second.m_Rotation, centerPoint) *
+
+				D2D1::Matrix3x2F::Skew(shape.second.m_Skew.x, shape.second.m_Skew.y, D2D1::Point2F(shape.second.m_Bounds.left, shape.second.m_Bounds.top)) *
+				D2D1::Matrix3x2F::Translation(shape.second.m_Offset) * D2D1::Matrix3x2F::Translation(D2D1::SizeF(GetMeterRectPadding().X, GetMeterRectPadding().Y)) *
+				D2D1::Matrix3x2F::Scale(shape.second.m_Scale, D2D1::Point2F(shape.second.m_Bounds.left, shape.second.m_Bounds.top))
+				);
+			canvas.DrawGeometry(shape.second, transform);
+		}
+	}
+	return true;
+}
+/*
+** Not necessary for the Geometry meter to be bound to any measures
+**
+*/
+void MeterGeometry::BindMeasures(ConfigParser & parser, const WCHAR * section)
```

Comment:
I didn't notice this until now... well there's another weird thing Visual Studio does (when auto generating methods into the cpp file)